### ISSUE TITLE
history.pushState is not available in packaged apps.

### ISSUE THREAD

**Issue by rpgobjects** [angular/angular.js#11932]
I get this error message when I run my chrome apps since m43 of chrome.  Seems related to this change. https://codereview.chromium.org/988293002

It doesn't have an negative effect on my apps so far, but I only get it when I use angular (3.15) in my chrome apps.

I saw a related closed issue.
https://github.com/angular/angular.js/issues/1358

I'm using the angular-csp.css file in my app.

**Comment by jdiamond** [comment]
This prevents the error message from appearing for me:

```
angular
    .module('app', [], function($provide) {
        // Prevent Angular from sniffing for the history API
        // since it's not supported in packaged apps.
        $provide.decorator('$window', function($delegate) {
            $delegate.history = null;
            return $delegate;
        });
    })
;
```

**Comment by caitp** [comment]
well judging from the history of this, it seems like this feature has been disabled in chrome platform apps for a very long time (the bug is about getting rid of an extended IDL attribute and making the bindings layer simpler --- was this working for you during some window in packaged apps before M43?)

At any rate, @jdiamond's decorator trick will work for now, so maybe we can close this? I don't think we have a reliable way of detecting whether we're running in crippled API mode or not

**Comment by raghavcreddy** [comment]
The decorator is not working anymore starting from Chrome 48.0.2564.97 (64-bit).

Is there any workaround?

**Comment by gkalpak** [comment]
It is strange, because this workaround shouldn't be affected by browser behavior.
Do you have a (preferrably minimal) working demo of the issue ?

**Comment by raghavcreddy** [comment]
Here's a minimal working demo of the issue. It's a Chrome packaged app.
https://github.com/raghavcreddy/ChromeAppAngular

**Comment by gkalpak** [comment]
Thx @raghavcreddy. I'll take a look.

**Comment by gkalpak** [comment]
@raghavcreddy, indeed `window.history` isn't writable in Chrome 48.
I tried defining a new `history` getter on `window` and it seemed to work (see below). It achieves the same effect as the original workaround.

```js
.config(function($provide) {
  $provide.decorator('$window', function($delegate) {
    Object.defineProperty($delegate, 'history', {get: () => null});
    return $delegate;
  });
});

// Or using the `.decorator()` helper:
.decorator('$window', function($delegate) {
  Object.defineProperty($delegate, 'history', {get: () => null});
  return $delegate;
});
```

**Comment by raghavcreddy** [comment]
Thank you @gkalpak

**Comment by gkalpak** [comment]
I have a fix for this is #13945. If that gets merged, then the workaround won't be necessary.
@raghavcreddy, btw, I think in Chrome Packaged Apps you should be running in [CSP mode][1].

[1]: https://docs.angularjs.org/api/ng/directive/ngCsp

**Comment by schellmax** [comment]
thanks a lot for sharing the workaround.
for those not familiar with ES6 yet (like me), here's the 'back-ported' version ;)
```
    .decorator('$window', function($delegate) {
        Object.defineProperty($delegate, 'history', {get: function(){ return null; }});
        return $delegate;
    })
```

**Comment by gkalpak** [comment]
BTW, if someone could try a (preferrably non-trivial Chrome Packaged App) with the files from [this build][1] and let me know if it logs any errors and if everything works as expected, that would awesome !
(Make sure you do **NOT** include the workaround.)

[1]: https://ci.angularjs.org/view/AngularJS%20Dev/job/angular.js-gkalpak/92/artifact/build/

**Comment by schellmax** [comment]
@gkalpak i've removed the workaround in my project and replaced angular.js by the one from your build, works just fine, no more errors in chrome app

**Comment by gkalpak** [comment]
@schellmax, that's good to know. Thx for trying it out :+1:
All we need now is to get #13945 merged :)

**Comment by donkonstante** [comment]
@gkalpak I tried getting the files from that build but it's gone. I get error 404.

**Comment by gkalpak** [comment]
@donkonstante, yeah orlder builds get deleted after some time. You can always checkout #13945 and build it yourself if you need it.

(I'll go lobby to get #13945 merged :stuck_out_tongue:)

**Comment by gkalpak** [comment]
It's in master and will be included in the next 1.5.x release.
@donkonstante, until then, you can grab the latest code from https://code.angularjs.org/snapshot/.

**Comment by husin-sajjadi** [comment]
I have the same problem in angular 9
How to fix that in angular 2+?
Thank you
